rspace-bench/src/test/scala/coop/rchain/rspace/bench/serialization/ModelSerializerBench.scala

```diff
@@ -0,0 +1,175 @@
+    out
+  }
+
+  def roundTripMany[A](seq: Seq[A])(implicit s: Serialize2ByteBuffer[A]): Seq[A] =
+    seq.map(roundTrip(_))
+
+  import collection.immutable.Seq
+
+  import org.scalacheck._
+  import org.scalacheck.Arbitrary._
+
+  import coop.rchain.rspace.Serialize
+
+  implicit def arbitraryDatum[C, T](chan: C)(
+      implicit
+      arbT: Arbitrary[T],
+      serializeC: Serialize[C],
+      serializeT: Serialize[T]
+  ): Arbitrary[Datum[T]] =
+    Arbitrary(for {
+      t <- arbT.arbitrary
+      b <- Arbitrary.arbitrary[Boolean]
+    } yield Datum.create(chan, t, b))
+
+  def arbitraryWaitingContinuation[C, P, K](chans: List[C])(
+      implicit
+      arbP: Arbitrary[P],
+      arbK: Arbitrary[K],
+      serializeC: Serialize[C],
+      serializeP: Serialize[P],
+      serializeK: Serialize[K]
+  ): Arbitrary[WaitingContinuation[P, K]] =
+    Arbitrary(
+      for {
+        pats         <- Gen.containerOfN[List, P](chans.length, arbP.arbitrary)
+        continuation <- arbK.arbitrary
+        boolean      <- Arbitrary.arbitrary[Boolean]
+      } yield WaitingContinuation.create(chans, pats, continuation, boolean)
+    )
+
+  implicit def arbitraryGnat()(
+      implicit
+      serializeC: Serialize[Par],
+      serializeP: Serialize[BindPattern],
+      serializeA: Serialize[ListParWithRandom],
+      serializeK: Serialize[TaggedContinuation]
+  ): Arbitrary[TestGNAT] =
+    Arbitrary(Gen.sized { size =>
+      val constrainedSize = if (size > 1) size else 1
+      for {
+        chans <- Gen.containerOfN[List, Par](constrainedSize, Arbitrary.arbitrary[Par])
+        data <- Gen.nonEmptyContainerOf[List, Datum[ListParWithRandom]](
+                 arbitraryDatum[Par, ListParWithRandom](chans.head).arbitrary
+               )
+        wks <- Gen
+                .nonEmptyContainerOf[List, WaitingContinuation[BindPattern, TaggedContinuation]](
+                  arbitraryWaitingContinuation[Par, BindPattern, TaggedContinuation](chans).arbitrary
+                )
+      } yield GNAT(chans, data, wks)
+    })
+
+  import coop.rchain.rholang.interpreter.storage.implicits._
+
+  val initSeed = 123456789L
+
+  val gnat: TestGNAT = gnat(10)
+
+  def gnat(size: Int): TestGNAT =
+    arbitraryGnat.arbitrary.apply(Parameters.default.withSize(size), Seed(initSeed + size)).get
+
+  val heavyGnats = (1 to 10).map(i => gnat(i))
+}
+
+@BenchState(Scope.Benchmark)
+class ProtobufModelBenchState extends ModelSerializerBenchState {
+
+  import scodec.Codec
+  import scodec.bits.BitVector
+
+  import coop.rchain.rholang.interpreter.storage.implicits._
+  implicit val cg: Codec[TestGNAT] = codecGNAT(
+    serializePar.toCodec,
+    serializeBindPattern.toCodec,
+    serializePars.toCodec,
+    serializeTaggedContinuation.toCodec
+  )
+  import coop.rchain.shared.ByteVectorOps._
+
+  implicit def serializer = new Serialize2ByteBuffer[TestGNAT] {
+
+    override def encode(a: TestGNAT): ByteBuffer =
+      cg.encode(a).get.toByteVector.toDirectByteBuffer
```

Comment:
This is adding up to 2 copies of the serialized data (depending on whether the `ByteVector` under the returned `Attempt[BitVector]` was `compact` or not). This might be giving protobuf an unfair handicap in this benchmark. Same goes for the usage of Codec, and adapting the proto serialization to the new `Serialize2ByteBuffer` interface.

What should be done instead, is comparing the speed of `Serialize.encode` for proto with `Serialize2ByteBuffer.encode` for kryo. Both of them are minimal in terms of copying done and number of passes through the serialized structure for their respective serialization framweorks, so they should be used as the benchmark.

Reply:
I was merely trying to mimic what is happening in RSpace: https://github.com/rchain/rchain/blob/1d60b733e427a716e8b708de2e0fc78c0d5601fa/rspace/src/main/scala/coop/rchain/rspace/LMDBOps.scala#L87
Remember that this benchmark is in the `rspaceBench` project, so it benches what we have versus what we could have

Reply:
The way I see it, is what we're **really** benching is not protobuf vs kryo, but

`protobuf + scodec + 2 or 3 unnecessary buffer copies` - codename Protobuf
vs
`kryo` - codename Kryo

I appreciate that it's the performance of the whole stack that we care about and would very much like to get rid of both protobuf and scodec. BUT...

It could very much be the case, that Kryo is actually slower on our data structures, but the scodec + its buffers + protoc/scodec API incompatibility overhead is so huge, that we'd be better off switching the whole stack (the last part is what the benchmark suggests).

But it is very much possible to get rid of scodec and the multiple buffers **without** transitioning to Kryo (at least for GNAT serialization). If there are people on the team more disgusted by protobufs than I am, I'm fine changing that part of rspace for them ;) Heck, we could probably even use the current GNAT case class and convert it to protos similarly to how we do for ParMap and EMap.

I'm in the painful process of reimplementing protobuf (de)serialization so that it is stacksafe. Should we transition to Kryo, I'll have to redo substantial part of the work. I'm fine with that, but given our timelines, I'd like us to at least know, that Kryo wins with Protobuf **on level ground**. So, bench `Serialize.encode` vs `Serialize2ByteBuffer.encode` on the same Par-s, on the same Exprs, show Kryo's better - and I'm all good :)

If Protobuf turns out to be faster though, we should probably reconsider our options.

Reply:
> I'm in the painful process of reimplementing protobuf (de)serialization so that it is stacksafe. Should we transition to Kryo, I'll have to redo substantial part of the work. 

we would use protobuf still for internode communication so your work won't go to waste.